Initial CONTRIBUTING.md proposal

jonlawlor opened the pull request:
Providing contribution guidelines helps newcomers be more productive, and helps firm up the project goals.  For #25 

PTAL @kortschak @btracey @vladimir-ch (and of course anyone else who might be interested).

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### Reporting Bugs
+
+When you encounter a bug, please open an issue on the corresponding repo.  Start the issue title with the repo/subrepo name, like `stat/distmv: Issue Name`.  Be specific about the environment you encountered the bug in.  If you are able to write a test that reproduces the bug, please include it in the issue.  As a rule we keep all tests OK.
```
No capitals in issue titles please, unless it is an exported label being discussed by the issue (same with PRs).

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### Suggesting Enhancements
+
+If the scope of the enhancement is small, open an issue.  If it is large, such as suggesting a new repo, sub-repo, or interface refactoring, then please start a discussion on the gonum-dev list.
```
Add link to gonum-dev?

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### Code Contribution
+
+If it is possible to split a large pull into two or more smaller pull requests, please try to do so.  Pulls should include tests for any new code before merging.  It is ok to start a pull request on partially implemented code to get feedback, and see if your approach to a problem is sound.  You don't need have to have tests, or even have code that compiles to open a pull.  Benchmarks are optional ...
```
Please don't truncate "pull request" to "pull".

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### Code Review
+
+If you are a contributor, please be welcoming to new contributors.  [Here](http://sarah.thesharps.us/2014/09/01/the-gentle-art-of-patch-review/) is a good guide.  We follow the convention of requiring at least 1 LGTM before a merge.  If a submitter or reviewer wants, they can request additional review from others and more LGTMs before merge, which should be used when the code is tricky or could b ...
```
"which should be used" seems odd, but I cannot think of better wording.

jonlawlor replied:
"When code is tricky or controversial, submitters and reviewers can request additional review from others and more LGTMs before merge at their discretion."

jonlawlor replied:
Done as "When code is tricky or controversial, submitters and reviewers can request additional review from others and more LGTMs before merge"

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### What Can I Do to Help?
+
+If you are looking for some way to help the Gonum project, there are good places to start, depending on what you are comfortable with.  You can [search](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+user%3Agonum) for open issues in need of resolution.  You can improve documentation, or write longer examples.  You can add and improve tests.  You can improve performance, either by ...
```
s/write longer examples/improve examples/

jonlawlor commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+
+### Project Scope
+
+The purpose of the Gonum project is to provide general purpose numerical computing and data science libraries for the Go programming language.  The libraries should aim to provide building blocks for disciplinary work and advanced algorithms.  Code should be implemented in pure Go. Calls to assembly, C, Fortran, or other languages may be justified with performance considerations, but should be op ...
```
The spaces are now moot, there is a newline after each.

jonlawlor replied:
Done.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Here are the current repos for the gonum project.
+If code you want to contribute doesn't quite fit in any of them, then please start a discussion on the [mailing list](https://groups.google.com/forum/#!forum/gonum-dev).
+
+* [matrix](https://github.com/gonum/matrix) - Matrix packages for the Go language.
```
I think the package description don't have to follow the descriptions on GitHub. Especially "Go" and "the Go language" can be omitted here.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+If you are a new contributor, thank you!  Before your first merge, you will need to be added to the [CONTRIBUTORS](https://github.com/gonum/license/blob/master/CONTRIBUTORS) and [AUTHORS](https://github.com/gonum/license/blob/master/AUTHORS) file.
+Open a pull request adding yourself to them.
+All Gonum code follows the BSD license in the license document.
```
Replace license with a link like CONTRIBUTORS and AUTHORS above.

jonlawlor replied:
Done

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+If you are a new contributor, thank you!  Before your first merge, you will need to be added to the [CONTRIBUTORS](https://github.com/gonum/license/blob/master/CONTRIBUTORS) and [AUTHORS](https://github.com/gonum/license/blob/master/AUTHORS) file.
+Open a pull request adding yourself to them.
+All Gonum code follows the BSD license in the license document.
+We prefer that code contributions do not come with additional licensing.
```
I find these two sentences strange. Aren't they saying almost the same thing? Also, what are the criteria for allowing non-BSD license? And if we allow, then which licenses are ok? I would like to know myself.

btracey replied:
Here is the intent of that line:
I've added code in mathext and in stat that is originally a line-for-line translation of other BSD code. It thus includes the additional license information. All code that is added as such must be BSD. However, in general we would prefer that code is generated from scratch, and so does not need to come with such additional license information.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+If it is possible to split a large pull into two or more smaller pull requests, please try to do so.
+Pulls should include tests for any new code before merging.
+It is ok to start a pull request on partially implemented code to get feedback, and see if your approach to a problem is sound.
+You don't need have to have tests, or even have code that compiles to open a pull.
```
need have

jonlawlor replied:
Done

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Pulls should include tests for any new code before merging.
+It is ok to start a pull request on partially implemented code to get feedback, and see if your approach to a problem is sound.
+You don't need have to have tests, or even have code that compiles to open a pull.
+Benchmarks are optional for new features, but if you are submitting a pull justified by performance improvment, you will need benchmarks to measure the impact of your change, and the pull should include a report from [benchcmp](https://godoc.org/golang.org/x/tools/cmd/benchcmp) or [benchstat](https://github.com/rsc/benchstat).
```
improvment

btracey replied:
"or, preferably, benchstat"

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+Code should be implemented in pure Go.
+Calls to assembly, C, Fortran, or other languages may be justified with performance considerations, but should be opt-in for users.
+
+### The Gonum Repos
```
Use repositories instead of repos?

jonlawlor replied:
Done

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+## Getting Started
+
+Gonum discussion takes place on the [gonum-dev google group](https://groups.google.com/forum/#!forum/gonum-dev).
```
Probably silly given where this lives, but maybe also 

Code is at github.com/gonum/repo
Documentation at godoc.org/github.com/gonum/repo

kortschak replied:
Code is at github.com/gonum/\<repo\>.
Documentation is at godoc.org/github.com/gonum/\<repo\>.

jonlawlor replied:
Done

jonlawlor commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,90 @@
+#### Table of Contents
+
+[Getting Started](#getting-started)
+  * [Code of Conduct](#code-of-conduct)
```
Done, sorry.

jonlawlor commented:
I've made most changes.  Some comments have not been resolved, but all will be listed as outdated.  Here is a summary:

* Should it be `Gonum` or `gonum`?  We have the latter in the license, and the former shows up a lot too.
* What descriptions should the repositories have?  I just used the description listed by github, but it doesn't have to match and it is often redundant.
* Should the document lead off with something like:
```
##Getting Started

Gonum discussion takes place on the [gonum-dev google group](https://groups.google.com/forum/#!forum/gonum-dev). *Code is at github.com/gonum/repo Documentation at godoc.org/github.com/gonum/repo*

```

* Do these licensing requirements make sense:
```
All Gonum code follows the BSD license in the license document.
We prefer that code contributions do not come with additional licensing.
```

kortschak commented:
What is the intention of the Getting Started section? It could probably do with a lead in.

The licensing section could be changed to "Gonum is licensed under a BSD-style license found in the LICENSE document. We prefer that code contributions do not come with additional licensing."

Also, since this is a meta decision, I think we should have a reasonable quorum of acceptances before it is committed.

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+If you are looking for some way to help the Gonum project, there are good places to start, depending on what you are comfortable with.
+You can [search](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+user%3Agonum) for open issues in need of resolution.
+You can improve documentation, or improve examples.
```
Also we could really do with a person with webdev experience to help build the site.

jonlawlor replied:
Basic stuff like what you can see at [scipy.org](http://scipy.org/) can be done with github pages, and if that's enough then I can put a skeleton together.  It would be great to have executable examples someplace like play.gonum.org, but that is beyond me at the moment.

kortschak commented:
>What descriptions should the repositories have? I just used the description listed by github, but it doesn't have to match and it is often redundant.

I think we could reverse the information flow here. If the names are decided in this PR in a way that makes them consistent, then the names can be changed on the repos as needed.

btracey commented:
Sorry, I don't understand what you mean.

kortschak commented:
[This](https://github.com/gonum/license/pull/26#issuecomment-257197087)? I mean agree on the package descriptions in this PR and then update on the repos to match.

btracey commented:
Ah, thanks.

sbinet commented:
the content looks good, thanks for tackling this.
but, (you knew there was a "but" coming :P) wouldn't this be more natural or more visible to have this on the gonum.org landing page?
`gonum/license` isn't a repo that will attract the eye of many contributors, I am afraid.
(of course, having the content duplicated or synchronized here as well does make sense, though)

jonlawlor commented:
When repos have a CONTRIBUTING.md file there is a little banner that shows up when creating an issue or pull request, like:

![image](https://cloud.githubusercontent.com/assets/4468162/19771421/e63be566-9c30-11e6-9733-a3f2e31520d1.png)

I think the idea will be to replicate it across the various repos.

I think maybe we should rename `gonum/license` to `gonum/gonum` or something like that though.

jonlawlor commented:
I've moved the text below the Getting Started section elsewhere, and fixed a few links.  I think the only thing outstanding is (possibly) reworking the various repository descriptions.

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,116 @@
+  * [What Can I Do to Help?](#what-can-i-do-to-help)
+  * [Style](#style)
+
+## Getting Started
```
I'm unclear on what this is for. With the text as it is, this seems like a heading for a heading's sake. Would "Overview" or "The Project" or something like that be better (I don't like either of these overmuch).

jonlawlor replied:
I adapted the structure from atom's [contribuing.md](https://github.com/atom/atom/blob/master/CONTRIBUTING.md) - they call it "What should I know before I get started?"

In ours the first part covers basically "what", and the contributing section is "how."   How about moving the Working Together subsection to the top of the Contributing section, and making the "Project Goals" and "The Gonum Repositories" top level sections instead of subsections?

kortschak replied:
SGTM

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,116 @@
+The purpose of the Gonum project is to provide general purpose numerical computing and data science libraries for the Go programming language.
+The libraries should aim to provide building blocks for disciplinary work and advanced algorithms.
+Code should be implemented in pure Go.
+Calls to assembly, C, Fortran, or other languages may be justified with performance considerations, but should be opt-in for users.
```
This is not true for assembly. C and Fortran calls are opt-in. ASM calls are opt-out.

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,117 @@
+The libraries should aim to provide building blocks for disciplinary work and advanced algorithms.
+Code should be implemented in pure Go.
+Calls to C, Fortran, or other languages may be justified with performance considerations, but should be opt-in for users.
+Calls to assembly should be opt-out.
```
"Calls to assembly should be opt-out, if included."

jonlawlor replied:
Done

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,115 @@
+
+## Project Scope
+
+The purpose of the Gonum project is to provide general purpose numerical computing and data science libraries for the Go programming language.
```
One thing that has not been mentioned here is that a central part of our motivation has been in providing code that fosters good peer-review practice. So code in the project favours readability and explicitness over cleverness.

jonlawlor replied:
Done, as
> Code should favor readability and explicitness over cleverness. This should make code easier to review, and also should provide meaningful examples of numerical computing in Go. Where possible, the source of algorithms should be referenced in the comments.

kortschak replied:
LGTM. Thank you.

btracey replied:
Is more of a manifesto appropriate here? Probably not, but maybe

"Code should favor readability and explicitness over cleverness. This makes code easy to review and verify, not only at submission, but also for users who want to understand how their algorithms work. Where possible ..."

jonlawlor replied:
I think that flows better, done.

jonlawlor commented:
As @kortschak suggested, we should have more support before a merge.  I've [posted](https://groups.google.com/forum/#!topic/gonum-dev/Wbw9LfCOvUc) on the mailing list to see if anyone has any objections or would like anything reworked or clarified.

kortschak commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,118 @@
+
+Here are the current repositories for the gonum project.
+If code you want to contribute doesn't quite fit in any of them, then please start a discussion on the [mailing list](https://groups.google.com/forum/#!forum/gonum-dev).
+Code can be found at [github.com/gonum/](https://github.com/gonum/)&lt;repo&gt;, and documentation at godoc.org/github.com/gonum/&lt;repo&gt;.
```
In md, \<word\> can be spelled as \\\<word\\\> to avoid htmlification.

jonlawlor commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,118 @@
+We prefer that code contributions do not come with additional licensing.
+For exceptions, added code must also follow a BSD license.
+
+### Code Contribution
```
Done

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,123 @@
+If code you want to contribute doesn't quite fit in any of them, then please start a discussion on the [mailing list](https://groups.google.com/forum/#!forum/gonum-dev).
+Code can be found at [github.com/gonum/](https://github.com/gonum/)\<repo\>, and documentation at godoc.org/github.com/gonum/\<repo\>.
+
+* [matrix](https://github.com/gonum/matrix) - Matrix packages for the Go language.
```
[matrix] - Implements data containers for matrix types (Dense, Symmetric), and higher-level operations using them (SVD, Linear solve)

[stat] - Contains statistics routines for linear and multi-dimensional data, univariate and multivariate sampling distributions, and advanced sampling routines

[blas] - Native Go BLAS implementation for float64 and float32, and c-wrapper for external libraries

[lapack] - Native Go implementation of many LAPACK routines and c-wrapper for LAPACKE.

[integrate] - Routines for numeric evaluation of integrals (quadrature)

[diff] - Routines for computing derivatives

[graph] - Implements types and interfaces for graphs (networks), and implements higher-level routines (A-star, Kruskal)

[optimize] - Package for searching for the optimal value of functions

[mathext] - implements basic elementary functions not included in the Go standard library

jonlawlor replied:
Done

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,123 @@
+
+The Gonum project provides general purpose numerical computing and data science libraries for the Go programming language, and we would like you to join us in improving Gonum's quality and scope.
+This document is for anyone who is contributing or interested in contributing.
+
```
Something about "Questions about gonum or use of its libraries can be directed to [gonum-dev](link)"

jonlawlor replied:
Done

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,123 @@
+If it is possible to split a large pull request into two or more smaller pull requests, please try to do so.
+Pull requests should include tests for any new code before merging.
+It is ok to start a pull request on partially implemented code to get feedback, and see if your approach to a problem is sound.
+You don't need to have tests, or even have code that compiles to open a pull request.
```
"to open a pull request (though you will before submission!)"

jonlawlor replied:
Done

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,123 @@
+### Code Review
+
+If you are a contributor, please be welcoming to new contributors.  [Here](http://sarah.thesharps.us/2014/09/01/the-gentle-art-of-patch-review/) is a good guide.
+We follow the convention of requiring at least 1 LGTM before a merge.
```
"1 LGTM (looks good to me)"

btracey commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,123 @@
+If you are a contributor, please be welcoming to new contributors.  [Here](http://sarah.thesharps.us/2014/09/01/the-gentle-art-of-patch-review/) is a good guide.
+We follow the convention of requiring at least 1 LGTM before a merge.
+When code is tricky or controversial, submitters and reviewers can request additional review from others and more LGTMs before merge.
+You can ask for more review by saying "PTAL" in a comment in a pull request.
```
"PTAL" ("Please take a(nother) look")

jonlawlor replied:
as above

jonlawlor commented:
Currently there are no outstanding requests to modify CONTRIBUTING.md.  I'll give gonum-dev some time to provide feedback.

btracey commented:
This LGTM. Thank you

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+# [<img src="https://avatars1.githubusercontent.com/u/3771091?v=3&s=30">](https://github.com/gonum) Contribution Guidelines
+#### Introduction
```
A couple of post-LGTM suggestions.

How about removing this `#### Introduction` header, it's not really necessary and I don't like the jump from one `#` to four `####`.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+This document is for anyone who is contributing or interested in contributing.
+Questions about Gonum or the use of its libraries can be directed to [gonum-dev](https://groups.google.com/forum/#!forum/gonum-dev).
+
+#### Table of Contents
```
This could have `##`.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+## Project Scope
+
+The purpose of the Gonum project is to provide general purpose numerical computing and data science libraries for the Go programming language.
+The libraries should aim to provide building blocks for disciplinary work and advanced algorithms.
```
What does "disciplinary work" mean here?

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+
+The purpose of the Gonum project is to provide general purpose numerical computing and data science libraries for the Go programming language.
+The libraries should aim to provide building blocks for disciplinary work and advanced algorithms.
+Code should be implemented in pure Go.
```
From here on the paragraph looks like it belongs in Contributing section rather than Project Scope.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+If code you want to contribute doesn't quite fit in any of them, then please start a discussion on the [mailing list](https://groups.google.com/forum/#!forum/gonum-dev).
+Code can be found at [github.com/gonum/](https://github.com/gonum/)\<repo\>, and documentation at godoc.org/github.com/gonum/\<repo\>.
+
+* [matrix](https://github.com/gonum/matrix) — Implements data containers for matrix types (Dense, Symmetric), and higher-level operations using them (SVD, Linear solve)
```
How about `Data containers for various matrix types, and higher-level operations using them (factorizations, solution of linear systems)`

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+Code can be found at [github.com/gonum/](https://github.com/gonum/)\<repo\>, and documentation at godoc.org/github.com/gonum/\<repo\>.
+
+* [matrix](https://github.com/gonum/matrix) — Implements data containers for matrix types (Dense, Symmetric), and higher-level operations using them (SVD, Linear solve)
+* [plot](https://github.com/gonum/plot) — A repository for plotting and visualizing data
```
Plotting and visualization of data

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+
+* [matrix](https://github.com/gonum/matrix) — Implements data containers for matrix types (Dense, Symmetric), and higher-level operations using them (SVD, Linear solve)
+* [plot](https://github.com/gonum/plot) — A repository for plotting and visualizing data
+* [stat](https://github.com/gonum/stat) — Contains statistics routines for linear and multi-dimensional data, univariate and multivariate sampling distributions, and advanced sampling routines
```
~~`Contains`~~ `Statistical routines ...` ?

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [matrix](https://github.com/gonum/matrix) — Implements data containers for matrix types (Dense, Symmetric), and higher-level operations using them (SVD, Linear solve)
+* [plot](https://github.com/gonum/plot) — A repository for plotting and visualizing data
+* [stat](https://github.com/gonum/stat) — Contains statistics routines for linear and multi-dimensional data, univariate and multivariate sampling distributions, and advanced sampling routines
+* [blas](https://github.com/gonum/blas) — Native Go BLAS implementation for float64 and float32, and c-wrapper for external libraries
```
C wrapper? cgo wrapper?

jonlawlor replied:
I think it is cgo wrapper, because the wrapper is written using cgo.

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [plot](https://github.com/gonum/plot) — A repository for plotting and visualizing data
+* [stat](https://github.com/gonum/stat) — Contains statistics routines for linear and multi-dimensional data, univariate and multivariate sampling distributions, and advanced sampling routines
+* [blas](https://github.com/gonum/blas) — Native Go BLAS implementation for float64 and float32, and c-wrapper for external libraries
+* [lapack](https://github.com/gonum/lapack) — Native Go implementation of many LAPACK routines and c-wrapper for LAPACKE
```
C wrapper? cgo wrapper?

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [blas](https://github.com/gonum/blas) — Native Go BLAS implementation for float64 and float32, and c-wrapper for external libraries
+* [lapack](https://github.com/gonum/lapack) — Native Go implementation of many LAPACK routines and c-wrapper for LAPACKE
+* [floats](https://github.com/gonum/floats) — Operations on slices of floats
+* [integrate](https://github.com/gonum/integrate) — Routines for numeric evaluation of integrals (quadrature)
```
numerical

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [floats](https://github.com/gonum/floats) — Operations on slices of floats
+* [integrate](https://github.com/gonum/integrate) — Routines for numeric evaluation of integrals (quadrature)
+* [diff](https://github.com/gonum/diff) — Routines for computing derivatives
+* [graph](https://github.com/gonum/graph) — Implements types and interfaces for graphs (networks), and implements higher-level routines (A-star, Kruskal)
```
`Interfaces and types for graphs, and higher-level algorithms using them (e.g., graph traversal, path finding, network analysis)`

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [integrate](https://github.com/gonum/integrate) — Routines for numeric evaluation of integrals (quadrature)
+* [diff](https://github.com/gonum/diff) — Routines for computing derivatives
+* [graph](https://github.com/gonum/graph) — Implements types and interfaces for graphs (networks), and implements higher-level routines (A-star, Kruskal)
+* [optimize](https://github.com/gonum/optimize) — Package for searching for the optimal value of functions
```
`Algorithms for solving optimization problems`

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [graph](https://github.com/gonum/graph) — Implements types and interfaces for graphs (networks), and implements higher-level routines (A-star, Kruskal)
+* [optimize](https://github.com/gonum/optimize) — Package for searching for the optimal value of functions
+* [unit](https://github.com/gonum/unit) — Package for converting between scientific units
+* [mathext](https://github.com/gonum/mathext) — Implements basic elementary functions not included in the Go standard library
```
~~Implements~~

vladimir-ch commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,132 @@
+* [optimize](https://github.com/gonum/optimize) — Package for searching for the optimal value of functions
+* [unit](https://github.com/gonum/unit) — Package for converting between scientific units
+* [mathext](https://github.com/gonum/mathext) — Implements basic elementary functions not included in the Go standard library
+* [internal](https://github.com/gonum/internal) — Internal routines for the gonum project
```
Gonum

jonlawlor commented:
@vladimir-ch I'm going to open a new pull to address your comments.
